File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

bool visited[8]={};
int answer =0;

void dfs(int sleep, vector<vector<int>> &dungeons,int count){
    answer = max(answer,count);
    
    for(int i=0; i<dungeons.size(); i++){
        if(!visited[i] && sleep >= dungeons[i][0]){
            visited[i] = true;
            dfs(sleep-dungeons[i][1],dungeons,count+1);
            visited[i]=false;
        }
    }
    
}

int solution(int k, vector<vector<int>> dungeons) {
    dfs(k, dungeons, 0);
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    boolean[] visited;
    int answer;
    
    void dfs(int hp, int count, int[][] dungeons){
        answer = Math.max(answer, count);
        
        for(int i=0; i<dungeons.length; i++){
            if(!visited[i] && hp>=dungeons[i][0]){
                visited[i] = true;
                count++;
                dfs(hp-dungeons[i][1],count,dungeons);
                count--;
                visited[i] = false;
            }
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        answer = 0;
        visited = new boolean[dungeons.length];
        dfs(k,0,dungeons);
        return answer;
    }
}